src/inc/til/small_vector.h

```diff
@@ -0,0 +1,747 @@
+            }
+            return _data[off];
+        }
+
+        void clear() noexcept
+        {
+            std::destroy_n(begin(), _size);
+
+            if (_capacity != N)
+            {
+                delete[] _data;
+            }
+
+            _data = &_buffer[0];
+            _capacity = N;
+            _size = 0;
+        }
+
+        void resize(size_type newSize)
+        {
+            if (newSize < _size)
+            {
+                std::destroy_n(_data + newSize, _size - newSize);
+            }
+            else if (newSize > _size)
+            {
+                if (newSize > _capacity)
+                {
+                    _grow(newSize - _capacity);
+                }
+
+                std::uninitialized_value_construct_n(_data + _size, newSize - _size);
+            }
+
+            _size = newSize;
+        }
+
+        void resize(size_type newSize, const_reference value)
+        {
+            if (newSize < _size)
+            {
+                std::destroy_n(_data + newSize, _size - newSize);
+            }
+            else if (newSize > _size)
+            {
+                if (newSize > _capacity)
+                {
+                    _grow(newSize - _capacity);
+                }
+
+                std::uninitialized_fill_n(_data + _size, newSize - _size, value);
+            }
+
+            _size = newSize;
+        }
+
+        void shrink_to_fit()
+        {
+            if (_capacity == N || _size == _capacity)
+            {
+                return;
+            }
+
+            auto data = &_buffer[0];
+            if (_size > N)
+            {
+                data = new T[_size];
+            }
+
+            std::uninitialized_move_n(begin(), _size, data);
+            std::free(_data);
+
+            _data = data;
+            _capacity = _size;
+        }
+
+        void push_back(const T& value)
+        {
+            if (_size == _capacity)
+            {
+                _grow(1);
+            }
+
+            new (_data + _size) T(value);
+            _size++;
+        }
+
+        void push_back(T&& value)
+        {
+            if (_size == _capacity)
+            {
+                _grow(1);
+            }
+
+            new (_data + _size) T(std::move(value));
+            _size++;
+        }
+
+        template<typename... Args>
+        reference emplace_back(Args&&... args)
+        {
+            if (_size == _capacity)
+            {
+                _grow(1);
+            }
+
+            const auto it = new (_data + _size) T(std::forward<Args>(args)...);
+            _size++;
+            return *it;
+        }
+
+        iterator insert(const_iterator pos, size_type count, const T& value)
+        {
+            assert(pos >= begin() && pos <= end());
+
+            const auto off = pos - begin();
+            const auto newSize = _size + count;
+
+            if (newSize > _capacity)
+            {
+                _grow(count);

```

Comment:
we actually need to grow by the difference between _capacity and `newSize`, right? because if...

```
size = 5
cap = 10
```

and we want to insert `7`...

we only need to grow by `2` to get cap to `12`?